src/main/scala/com/exasol/cloudetl/kinesis/KinesisShardsMetadataReader.scala

```diff
@@ -0,0 +1,103 @@
+    val shards = getAllShardsFromStream(streamName, amazonKinesis)
+    val shardsMetadataFromStream = collectShardsIds(shards)
+    val kinesisMetadata =
+      combineShardsMetadata(shardsMetadataFromTable, shardsMetadataFromStream)
+    try {
+      kinesisMetadata.foreach {
+        case (shardId, sequenceNumber) => exaIterator.emit(shardId, sequenceNumber)
+      }
+    } catch {
+      case exception @ (_: ExaDataTypeException | _: ExaIterationException) =>
+        val message = exception.getMessage
+        throw new KinesisConnectorException(
+          s"KinesisShardsMetadataReader cannot emit shards' ids. Caused: $message",
+          exception
+        )
+    } finally {
+      amazonKinesis.shutdown()
+    }
+  }
+
+  private[kinesis] def getShardsMetadataFromTable(
+    exaIterator: ExaIterator
+  ): mutable.Map[String, String] = {
+    val shardsMetadata: mutable.HashMap[String, String] = mutable.HashMap.empty[String, String]
+    do {
+      val shardId = exaIterator.getString(SHARD_ID_INDEX)
+      val sequenceNumber = exaIterator.getString(SEQUENCE_NUMBER_INDEX)
+      shardsMetadata += (shardId -> sequenceNumber)
+    } while (exaIterator.next())
+    shardsMetadata
+  }
+
+  // Obtains shards from a Stream until there is no more shards.
+  // https://docs.aws.amazon.com/streams/latest/dev/kinesis-using-sdk-java-retrieve-shards.html
+  private[kinesis] def getAllShardsFromStream(
+    streamName: String,
+    amazonKinesis: AmazonKinesis
+  ): List[Shard] = {
+    val describeStreamRequest = new DescribeStreamRequest
+    describeStreamRequest.setStreamName(streamName)
+    val shards = new util.ArrayList[Shard]
+    var exclusiveStartShardId: String = null
+    do {
+      describeStreamRequest.setExclusiveStartShardId(exclusiveStartShardId)
+      val describeStreamResult = amazonKinesis.describeStream(describeStreamRequest)
+      val _ = shards.addAll(describeStreamResult.getStreamDescription.getShards)
+      if (describeStreamResult.getStreamDescription.getHasMoreShards && !shards.isEmpty) {
+        exclusiveStartShardId = shards.get(shards.size - 1).getShardId
+      } else {
+        exclusiveStartShardId = null
+      }
+    } while ({
+      exclusiveStartShardId != null
+    })
+    shards.asScala.toList
+  }
+
+  private[kinesis] def collectShardsIds(shards: List[Shard]): List[String] =
+    shards.toStream.map(shard => shard.getShardId).toList
+
+  private[kinesis] def combineShardsMetadata(
+    shardsMetadataFromTable: mutable.Map[String, String],
+    shardsMetadataFromStream: List[String]
+  ): mutable.HashMap[String, String] = {
+    val kinesisMetadata: mutable.HashMap[String, String] = mutable.HashMap.empty[String, String]
+    shardsMetadataFromStream.foreach(shardId => {
+      val sequenceNumber: String = shardsMetadataFromTable.getOrElse(shardId, null)
+      kinesisMetadata += (shardId -> sequenceNumber)
+    })
+    kinesisMetadata
```

Comment:
Same as above. Accept immutable parameter and return immutable map.
Function local data structure can be mutable.

Reply:
done